File: steps/5/main/scala/eightyDays.scala/Bank.scala
package eightyDays.scala

case class Bank(name: String) {
  private[this] var partners = Map[Identification, Partner]()

  def partner(id: Identification) = partners get id

  def addPartner(pPartner: Partner) = partners
                                        .find(_._2 == pPartner)
                                        .map(_._1)
                                        .getOrElse {
                                          val newId = Identification()
                                          partners += (newId -> pPartner)
                                          newId
                                        }
}

File: steps/5/main/java/eightyDays/java8/Bank.java
package eightyDays.java8;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public class Bank {
    private final String name;
    private Map<Identification, Partner> partners = new HashMap<>();

    public Bank(String pName) {
        name = pName;
    }

    public String getName() {
        return name;
    }

    public Optional<Partner> getPartner(Identification id) {
        return Optional.ofNullable(partners.get(id));
    }

    public Identification addPartner(Partner pPartner) {
        return partners
                .entrySet()
                .stream()
                .filter(entry -> entry.getValue().equals(pPartner))
                .findFirst()
                .map(Map.Entry::getKey)
                .orElseGet(() -> {
                    Identification newId = new Identification();
                    partners.put(newId, pPartner);
                    return newId;
                });
    }
}
